### PR TITLE
Surface Mesh Simplification: Introduce placement that bounds change of face normal

### LINKED ISSUE #2192: Surface_mesh_simplification: simplification leading to self intersections.
## Issue Details

The placement classes of the package may lead to self intersections.
sms-selfintersection

## Source Code

The polyhedron demo produced the above.

### PR DESCRIPTION
**Pull request by afabri**
## Summary of Changes

We add a placement `Bounded_normal_change_placement`  that is a wrapper around another placement which tests that the normals of the stars of the two vertices of the edge to collapse, does not differ by more than 90 degree before and after the collapse. If it differs too much it returns `boost::none`.

This PR adds a class, its reference manual, an example, and a section in the manual.

## Release Management

* Affected package(s): Surface_mesh_simplification
* Issue(s) solved (if any): fix #2192 
* Feature/Small Feature (if any): [Features/Small Features/Mesh Simplification: Add a placement][sf]

  [sf]: https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Mesh_Simplification:_Add_a_placement

### CODE REVIEW

File: Surface_mesh_simplification/doc/Surface_mesh_simplification/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/Bounded_normal_change_placement.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
+
```
Comment (sloriot) [inline Surface_mesh_simplification/doc/Surface_mesh_simplification/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/Bounded_normal_change_placement.h]: License header + package license include directive missing

File: Surface_mesh_simplification/doc/Surface_mesh_simplification/Surface_mesh_simplification.txt [hunk 2]
```diff
@@ -301,13 +301,33 @@ What is particular in this example is the property map that allows to associate
...
+
+
+\cgalFigureBegin{SMS-selfintersection, SMS-selfintersection.png}
+Simple mesh before and after the collapse of edge `v-w` into vertex `w`. While the normals of `f1`and `f2` are almost equal, they are opposed after the edge collapse.
+\cgalFigureEnd
+
+The class `Bounded_normal_change_placement` is a placement that checks if
+another placement would invert the normal of a face around the stars of the
+two vertices of an edge that is candidate for an edge collapse.  It then
+returns
```
Comment (sloriot) [inline Surface_mesh_simplification/doc/Surface_mesh_simplification/Surface_mesh_simplification.txt]: missing the end of the sentence.

File: Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/Bounded_normal_change_placement.h [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+#include <CGAL/license/Surface_mesh_simplification.h>
+
+
+namespace CGAL {
+
+namespace Surface_mesh_simplification  
+{
+
+  template<class Placement>
+  class Bounded_normal_change_placement
```
Comment (sloriot) [inline Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/Bounded_normal_change_placement.h]: indentation issue

File: Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/Bounded_normal_change_placement.h [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  
+  Bounded_normal_change_placement(const Placement& placement = Placement() )
+    : mPlacement(placement)
+  {}
+     
+  template <typename Profile> 
+  optional<typename Profile::Point>
+  operator()( Profile const& aProfile) const
+  {
+    optional<typename Profile::Point> op = mPlacement(aProfile);
```
Comment (MaelRL) [inline Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/Bounded_normal_change_placement.h]: Maybe add a comment over there that explains what triangles() does (first two triangles are incident to the edge about to be collapsed, point being moved is always in second position)

File: Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/Bounded_normal_change_placement.h [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    optional<typename Profile::Point> op = mPlacement(aProfile); 
+    if(op){
+       const typename Profile::Triangle_vector& triangles = aProfile.triangles();
+       if(triangles.size()>2){
+         typedef typename Profile::Point Point;
+         typedef typename Profile::Kernel Traits;
+         typedef typename Traits::Vector_3 Vector;
+         typename Profile::VertexPointMap ppmap = aProfile.vertex_point_map();
+         typename Profile::Triangle_vector::const_iterator it = triangles.begin();
+         ++it; ++it;
```
Comment (MaelRL) [inline Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/Bounded_normal_change_placement.h]: If the edge being collapsed is on the border, this will be wrong. Or can we never call this `Profile` with a border collapse ?

File: Surface_mesh_simplification/doc/Surface_mesh_simplification/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/Bounded_normal_change_placement.h [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+namespace CGAL {
+namespace Surface_mesh_simplification {
+
+/*!
+\ingroup PkgSurfaceMeshSimplification
+
+The class `Bounded_normal_change_placement` is a model for the `GetPlacement` concept
+which serves as a filter for another placement. It rejects the placement, if any
```
Comment (MaelRL) [inline Surface_mesh_simplification/doc/Surface_mesh_simplification/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/Bounded_normal_change_placement.h]: No need for the "," before if

**Comment by lrineau** [comment]
The was first errors in Travis:
1. Compilation error:
```
[ 42%] Building CXX object CMakeFiles/edge_collapse_polyhedron.dir/edge_collapse_polyhedron.cpp.o
/home/travis/build/CGAL/cgal/Surface_mesh_simplification/examples/Surface_mesh_simplification/edge_collapse_surface_mesh.cpp:137:51: error: 
      no member named 'Edge_length_cost' in namespace
      'CGAL::Surface_mesh_simplification'
            ,CGAL::parameters::get_cost     (SMS::Edge_length_cost  <Sur...
                                             ~~~~~^
```
https://travis-ci.org/CGAL/cgal/jobs/245685275#L4698

2. Bad permissions
```
.. Testing permissions...
Surface_mesh_simplification/doc/Surface_mesh_simplification/fig/SMS-selfintersection.png
Surface_mesh_simplification/examples/Surface_mesh_simplification/data/fold.off
```
https://travis-ci.org/CGAL/cgal/jobs/245685197#L4346

And then:

3. Now there is also a merge conflict with `master`.

**Comment by lrineau** [comment]
The [small feature page][1] is almost empty of content.

  [1]: https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Mesh_Simplification:_Add_a_placement

What is missing:
  - submission date,
  - the submission itself, with the detailed changes, and links to the modified or new documentation pages, in the [First Round][2] section.

  [2]: https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Mesh_Simplification:_Add_a_placement#Submission

File: Surface_mesh_simplification/doc/Surface_mesh_simplification/Surface_mesh_simplification.txt [hunk 7]
```diff
@@ -301,13 +301,33 @@ What is particular in this example is the property map that allows to associate
...
+
+\subsection Surface_mesh_simplificationExamplewithBoundedNormal Example with Bounded Changement of Face Normals
+
+The surface mesh simplification does not guarantee that the resulting surface
+has no self intersections. Even the rather trivial mesh shown in \cgalFigureRef{SMS-selfintersection}
+results in a self intersection when one edge is collapsed using the Lindstrom-Turk method.
+
+
+\cgalFigureBegin{SMS-selfintersection, SMS-selfintersection.png}
+Simple mesh before and after the collapse of edge `v-w` into vertex `w`. While the normals of `f1`and `f2` are almost equal, they are opposed after the edge collapse.
```
Comment (sloriot) [inline Surface_mesh_simplification/doc/Surface_mesh_simplification/Surface_mesh_simplification.txt]: missing space before and

Comment (afabri) [inline Surface_mesh_simplification/doc/Surface_mesh_simplification/Surface_mesh_simplification.txt]: fixed

File: Surface_mesh_simplification/doc/Surface_mesh_simplification/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/Bounded_normal_change_placement.h (hunk 6 above)
Comment (afabri) [inline Surface_mesh_simplification/doc/Surface_mesh_simplification/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/Bounded_normal_change_placement.h]: fixed

File: Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/Bounded_normal_change_placement.h (hunk 3 above)
Comment (afabri) [inline Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/Bounded_normal_change_placement.h]: fixed

File: Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/Bounded_normal_change_placement.h (hunk 4 above)
Comment (afabri) [inline Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/Bounded_normal_change_placement.h]: I added 
`/* triangles returns the triangles of the star of the vertices of the edge to collapse. 
 First the two trianges incident to the edge, then the other triangles. 
 The second vertex of each triangle is the vertex that gets placed */`
